wallet/wallet.c

```diff
@@ -595,6 +595,46 @@ wallet_htlc_sigs_load(const tal_t *ctx, struct wallet *w, u64 channelid)
 	return htlc_sigs;
 }
 
+bool wallet_remote_ann_sigs_load(const tal_t *ctx, struct wallet *w, u64 id,
+				 secp256k1_ecdsa_signature **remote_ann_node_sig,
+				 secp256k1_ecdsa_signature **remote_ann_bitcoin_sig)
+{
+	sqlite3_stmt *stmt;
+	int res;
+	stmt = db_select_prepare(w->db,
+				 "remote_ann_node_sig, remote_ann_bitcoin_sig"
+				 " FROM channels WHERE id = ?");
+	sqlite3_bind_int64(stmt, 1, id);
+
+	res = sqlite3_step(stmt);
+
+	/* This must succeed, since we know the channel exists */
+	assert(res == SQLITE_ROW);
+
+	*remote_ann_node_sig = *remote_ann_bitcoin_sig = NULL;
+
+	if (sqlite3_column_type(stmt, 0) != SQLITE_NULL) {
+		*remote_ann_node_sig = tal(ctx, secp256k1_ecdsa_signature);
+		if (!sqlite3_column_signature(stmt, 0, *remote_ann_node_sig))
+			goto fail;
+	}
+
+	if (sqlite3_column_type(stmt, 1) != SQLITE_NULL) {
+		*remote_ann_bitcoin_sig = tal(ctx, secp256k1_ecdsa_signature);
+		if (!sqlite3_column_signature(stmt, 1, *remote_ann_bitcoin_sig))
+			goto fail;
+	}
+
+	db_stmt_done(stmt);
+	return true;
+
+fail:
+	*remote_ann_node_sig = tal_free(*remote_ann_node_sig);
+	*remote_ann_bitcoin_sig = tal_free(*remote_ann_bitcoin_sig);
```

Comment:
This can easily lead to a double-free or freeing a random memory location since we do not allocate in all cases. It'd be better to check both fields for `SQLITE_NULL` at the top, and set the out-args to `NULL` and return immediately if either is `SQLITE_NULL`. If both are not `SQLITE_NULL` initialized both using `tal` and then `goto fail;` if either of the `sqlite3_column_signature` calls fails. That way we either initialize all or none, and free only when both are actually initialized.

Reply:
Thank you! What you said is better, kinds of cases make sense easily.